docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,416 @@
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task resource.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the EtcdOperatorTask resource.
+  Spec EtcdOperatorTaskSpec `json:"spec"`
+  // Status is most recently observed status of the EtcdOperatorTask resource.
+  Status EtcdOperatorTaskStatus `json:"status,omitempty"`
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `.spec.type` field should be defined. It can have values from all supported `out-of-band` tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `.spec.config` field should be defined of type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task can have different input configuration.
+
+```go
+// EtcdOperatorTaskSpec is the spec for a EtcdOperatorTask resource.
+type EtcdOperatorTaskSpec struct {
+  
+  // Type specifies the type of out-of-band operator task to be performed. 
+  Type string `json:"type"`
+
+  // Config is a task specific configuration.
+  Config *runtime.RawExtension `json:"config,omitempty"`
+
+  // TTLSecondsAfterFinished is the time-to-live to garbage collect the 
+  // related resource(s) of task once it has been completed.
+  // +optional
+  TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+
+  // OwnerEtcdReference refers to the name and namespace of the corresponding 
+  // etcd owner for which the task has been invoked.
+  OwnerEtcdRefrence metav1.ObjectMeta `json:"ownerEtcdRefrence"`
+}
+```
+
+#### Status
+
+The authors propose that the following fields should be specified in the `Status` (current state) of the `EtcdOperatorTask` custom resource as the custom resource's `Status` will be used to monitor the progress of the task.
+
+* To capture the current state of a task, a field `.status.taskState` is defined in status.
+* If an operation involves many stages, `.status.lastOperation` will be useful in capturing the status of any intermediate stage.
+
+```go
+// TaskState represents the state of the task.
+type TaskState string
+
+// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
+type EtcdOperatorTaskStatus struct {
+  // ObservedGeneration is the most recent generation observed for the resource.
+  ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+  // State of the task is the last known state of the task.
+  State TaskState `json:"taskState"`
```

Comment:
Didn't we decide that we didn't want this to be called `State`, since it will end up looking like `Status.State`, which doesn't sound very good. Why don't we go with `Phase` instead?
cc @unmarshall

Reply:
I can rename it to `CurrentState` if you like but `Phase` doesn't seem correct to me 🤔

Reply:
I have renamed it to `CurrentState`. PTAL

Reply:
I don't think CurrentState is a standard Kubernetes way of defining the status field which indicates the state of the resource/operation. The word `current` becomes redundant, since the status anyway reflects the current state of the operation, and doesn't reflect the "stale"/"previous" state of the operation on purpose.

From the [k8s API conventions document](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties), you can actually see that status fields that indicate the state/phase of the resource, should be preferably modelled as conditions instead.

@ishan16696 @unmarshall do you think this makes sense for `Task` resource?

Reply:
```yaml
status:
  phase: rejected/failed
```
this doesn't seem good to me 🤔 , I prefer `taskState` or `currentState`

Reply:
We (Shreyas and I) discussed and we cannot come with anything better. So just stick to `state` for now. If something strikes us later then we can change it then.

Reply:
@ishan16696 can you please change it back from `CurrentState` to `State`? (both the field name and json tag)

Reply:
already made the change, just I haven't pushed it yet as I'm waiting for other comments to resolve.